### PR TITLE
Optimize elliptic solver (3): Sparse subdomain data

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

At some point the elliptic solver builds a matrix representation of a Poisson operator by feeding it unit vectors, so most of the data is actually zero (i.e. sparse). This PR makes the DG operator a lot faster when it operates on sparse data.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp [hunk 1]
```diff
@@ -273,6 +277,8 @@ struct DgOperatorImpl<System, Linearized, tmpl::list<PrimalFields...>,
       const std::tuple<FluxesArgs...>& fluxes_args,
       const std::tuple<SourcesArgs...>& sources_args,
       const DirectionMap<Dim, std::tuple<FluxesArgs...>>& fluxes_args_on_faces,
+      const std::function<bool(const ElementId<Dim>&)>& data_is_zero =
```
Comment (nilsdeppe) [inline src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp]: I haven't looked at any of the code in detail. Note that `std::function` is pretty slow compared to lambdas or a template parameter. If the  `std::function` is needed in a loop you might want to think about an alternative way of handling this. Basically, `std::function` does type erasure internally, which is why it's slower

Comment (wthrowe) [reply src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp]: I think making the argument type a template argument will work without having to change anything else.  (And all this appears again below.)

File: src/Elliptic/DiscontinuousGalerkin/SubdomainOperator/SubdomainOperator.hpp [hunk 2]
```diff
@@ -512,6 +557,12 @@ struct SubdomainOperator
                 }
               }
             }
+            ASSERT(
+                false,
```
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/SubdomainOperator/SubdomainOperator.hpp]: `ERROR`?  Evaluating the condition is free, no reason to skip it.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp (hunk 1 above)
Comment (nilsvu) [reply src/Elliptic/DiscontinuousGalerkin/DgOperator.hpp]: Changed in a fixup, please take a look

File: src/Elliptic/DiscontinuousGalerkin/SubdomainOperator/SubdomainOperator.hpp [hunk 3]
```diff
@@ -557,12 +557,10 @@ struct SubdomainOperator
...
-                false,
-                "The neighbor's neighbor "
-                    << neighbors_neighbor_id
-                    << " is part of the subdomain, but we didn't find its "
-                       "overlap ID. This is a bug, so please file an issue.");
+            ERROR("The neighbor's neighbor "
+                  << neighbors_neighbor_id
+                  << " is part of the subdomain, but we didn't find its "
+                     "overlap ID. This is a bug, so please file an issue.");
             return std::nullopt;
```
Comment (wthrowe) [inline src/Elliptic/DiscontinuousGalerkin/SubdomainOperator/SubdomainOperator.hpp]: Remove statement after `ERROR`.

**Review by wthrowe — CHANGES_REQUESTED** [review]
You can squash this change directly.

**Comment by nilsvu** [comment]
@wthrowe pushed another fixup to avoid timeouts in unit tests.

@nikwit Could you do the second review here?

**Comment by wthrowe** [comment]
Timeout fix looks good.  Squash.

**Review by wthrowe — APPROVED** [review]

**Review by nikwit — APPROVED** [review]
LGTM

**Pull request merged**